### PR TITLE
onchain-development anchor-pdas updated

### PR DESCRIPTION
**Pull request by SAMAD101**
### Summary of Changes

- Updated the Lesson section
- Updated the PDA section with more elaborative and clear language
- Updates to the different code snippets
- updated with the latest version of Anchor
- added some more explanation to some topics for a better understanding, overall improved the quality of the doc.

### CODE REVIEW

File: content/courses/onchain-development/anchor-pdas.md [hunk 1]
```diff
@@ -24,16 +24,14 @@ In this lesson you'll learn how to work with PDAs, reallocate accounts, and
...
 
-In addition to using constraints for account validation, some constraints can
-handle repeatable tasks that would otherwise require a lot of boilerplate inside
-our instruction logic. This lesson will introduce the `seeds`, `bump`,
-`realloc`, and `close` constraints to help you initialize and validate PDAs,
-reallocate accounts, and close accounts.
+In addition to validating accounts, some constraints can automate tasks that
+would otherwise require repetitive code in our instructions. This lesson will
+cover the `seeds`, `bump`, `realloc`, and `close` constraints to help you easily
+handle PDAs, reallocate space, and close accounts.
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:34]: I can already tell here this PR is going to be a massive improvement. The writing style is much more specific and to-the-point.👍

File: content/courses/onchain-development/anchor-pdas.md [hunk 2]
```diff
@@ -110,7 +116,7 @@ pub struct InitializePda<'info> {
         seeds = [b"example_seed", user.key().as_ref()],
         bump,
         payer = user,
-        space = 8 + 8
+        space = (DISCRIMINATOR as usize) + Accountype::INIT_SPACE
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:119]: We normally just make DISCRIMINATOR be a usize in the first place (than than a u8 and then coercing it). What do you think?

File: content/courses/onchain-development/anchor-pdas.md [hunk 3]
```diff
@@ -310,42 +312,45 @@ pub struct ReallocExample<'info> {
...
 
-Notice that `realloc` is set to `8 + 4 + instruction_data.len()`. This breaks
-down as follows:
+The `realloc` constraint from the above example can be broken down as follows:
 
-- `8` is for the account discriminator
-- `4` is for the 4 bytes of space that BORSH uses to store the length of the
-  string
+- the discriminator (8 bytes)
+- the length of the Borsh serialized data (4 bytes)
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:327]: The previous version was better here - this isn't the data, it's space used to store the length.

File: content/courses/onchain-development/anchor-pdas.md [hunk 4]
```diff
@@ -448,40 +453,27 @@ pub mod anchor_movie_review_program {
 }
 
 #[account]
+#[derive(InitSpace)]
 pub struct MovieAccountState {
     pub reviewer: Pubkey,    // 32
     pub rating: u8,          // 1
+    #[max_len(20)]
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:460]: Question: does max_len include the 4 bytes? So the string can only be 16 bytes long? Or does it exclude the length storage, so the string can be 20 bytes long? The Anchor docs are vague about it.

**Comment by mikemaccana** [comment]
Near perfect. A couple of minor bits.

Also please update https://github.com/Unboxed-Software/anchor-movie-review-program to be https://github.com/solana-developers/anchor-movie-review-program and https://github.com/Unboxed-Software/anchor-student-intro-program to https://github.com/solana-developers/anchor-student-intro-program (these repos are already updated)

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Minor changes needed (see above) but nearly done.

File: content/courses/onchain-development/anchor-pdas.md (hunk 1 above)
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:34]: thanks

File: content/courses/onchain-development/anchor-pdas.md (hunk 3 above)
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:327]: sure, I'll revert the changes in this part.

File: content/courses/onchain-development/anchor-pdas.md (hunk 2 above)
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:119]: DISCRIMINATOR is u8 in Anchor so I thought that making the constant also u8 might be a good idea. That sort of would be a good way to relate ig...
well we can make DISCRIMINATOR usize in the lesson, not a problem either.

**Comment by SAMAD101** [comment]
> Near perfect. A couple of minor bits.
> 
> Also please update https://github.com/Unboxed-Software/anchor-movie-review-program to be https://github.com/solana-developers/anchor-movie-review-program and https://github.com/Unboxed-Software/anchor-student-intro-program to https://github.com/solana-developers/anchor-student-intro-program (these repos are already updated)

but those pages are "Not found". 
Do I have to update the sources as well ?

File: content/courses/onchain-development/anchor-pdas.md (hunk 4 above)
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:461]: I suppose its added separately keeping the string 20 chars long. 
As string are actually `Vec<u8>` when serialized, having the length prefix (4 bytes) and the actual string length (20 bytes).

Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:461]: and the `#[derive(InitSpace)]` deals with all the space allocation in this case.

File: content/courses/onchain-development/anchor-pdas.md (hunk 2 above)
Comment (jereldlimjy) [inline content/courses/onchain-development/anchor-pdas.md:119]:
```suggestion
        space = (DISCRIMINATOR as usize) + AccountType::INIT_SPACE
```

File: content/courses/onchain-development/anchor-pdas.md [hunk 5]
```diff
@@ -54,31 +52,38 @@ struct ExampleAccounts {
...
-`pda_account` and `user`. The `pda_account` is constrained such that the seeds
-must be the string "example_seed," the public key of `user`, and the string
-passed into the instruction as `instruction_data`.
+For example, the following example shows a list of accounts that include:
+
+- `pda_account`
+- `user`
+
+The `pda_account` is constrained such that the seeds must be the string
+"example_seed," the public key of `user`, and the string passed into the
```
Comment (jereldlimjy) [inline content/courses/onchain-development/anchor-pdas.md:74]:
```suggestion
"example_seed", the public key of `user`, and the string passed into the
```

File: content/courses/onchain-development/anchor-pdas.md [hunk 6]
```diff
@@ -54,31 +52,38 @@ struct ExampleAccounts {
...
-the `seeds` constraint and verify that the account passed into the instruction
-matches the PDA found using the specified `seeds`.
+During account validation, Anchor will use the specified seeds to derive a PDA
+and check if the provided account matches the derived PDA.
 
 When the `bump` constraint is included without specifying a specific bump,
-Anchor will default to using the canonical bump (the first bump that results in
-a valid PDA). In most cases you should use the canonical bump.
+Anchor will use the canonical bump (the first bump that results in a valid PDA,
+with a value of 255). Typically, you should use the canonical bump.
```
[the remark this answers is not in the document]
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:60]: yeah, Anchor first tries to find the PDA using the canonical bump (255), if not found; then Anchor decrements it by one (upto 0) until it finds the PDA.

[the remark this answers is not in the document]
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md:60]: oh yes, my bad
canonical bump is just the first bump that results in a PDA; it doesn't have to be 255.
Thanks for getting me out of this confusion.
I'll update the doc.

File: content/courses/onchain-development/anchor-pdas.md [hunk 7]
```diff
@@ -310,42 +312,46 @@ pub struct ReallocExample<'info> {
 }
 
 #[account]
+#[derive(InitSpace)]
 pub struct AccountType {
     pub data: String,
 }
+
+const DISCRIMINATOR: u8 = 8;
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:320]: Why not just use `usize` for the DISCRIMINATOR and then not have to do the `as` conversion above?

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Just needs the `usize` change and we're ready to merge.

**Comment by SAMAD101** [comment]
> Just needs the `usize` change and we're ready to merge.

Done !

File: content/courses/onchain-development/anchor-pdas.md [hunk 8]
```diff
@@ -310,42 +312,46 @@ pub struct ReallocExample<'info> {
 }
 
 #[account]
+#[derive(InitSpace)]
 pub struct AccountType {
     pub data: String,
 }
+
+const DISCRIMINATOR: usize = 8;
+const BORSH_LEN: usize = 4;
```
Comment (mikemaccana) [inline content/courses/onchain-development/anchor-pdas.md:321]: One last change - often descriptions or comments give us a good hint for what a better name could be:

>  the `BORSH_LEN` is `4` for the space required to store the length of the
  string
  
Rename `BORSH_LEN` to `STRING_SIZE_SPACE` or similar.

**Review by mikemaccana — APPROVED** [review]
Approved, but please make the one small change above.

File: content/courses/onchain-development/anchor-pdas.md (hunk 8 above)
Comment (SAMAD101) [reply content/courses/onchain-development/anchor-pdas.md]: done

**Pull request merged**